File: java-schema-sdk/src/main/java/org/openmhealth/schema/domain/omh/DescriptiveStatistic.java
/*
 * Copyright 2015 Open mHealth
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.openmhealth.schema.domain.omh;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import javax.annotation.Nullable;
import java.util.HashMap;
import java.util.Map;


/**
 * A descriptive statistic of a set of measurements.
 *
 * @author Emerson Farrugia
 * @version 1.0
 * @see
 * <a href="http://www.openmhealth.org/documentation/#/schema-docs/schema-library/schemas/omh_descriptive-statistic">descriptive-statistic</a>
 */
public enum DescriptiveStatistic implements SchemaEnumValue, SchemaSupport {

    AVERAGE,
    MAXIMUM,
    MINIMUM,
    STANDARD_DEVIATION,
    VARIANCE,
    SUM,
    MEDIAN;

    public static final SchemaId SCHEMA_ID = new SchemaId(OMH_NAMESPACE, "descriptive-statistic", "1.0");

    private String schemaValue;
    private static final Map<String, DescriptiveStatistic> constantsBySchemaValue = new HashMap<>();

    static {
        for (DescriptiveStatistic constant : values()) {
            constantsBySchemaValue.put(constant.getSchemaValue(), constant);
        }
    }

    DescriptiveStatistic() {
        this.schemaValue = name().toLowerCase().replace('_', ' ');
    }

    @Override
    public SchemaId getSchemaId() {
        return SCHEMA_ID;
    }

    @Override
    @JsonValue
    public String getSchemaValue() {
        return this.schemaValue;
    }

    @Nullable
    @JsonCreator
    public static DescriptiveStatistic findBySchemaValue(String schemaValue) {
        return constantsBySchemaValue.get(schemaValue);
    }
}

File: kotlin-schema-sdk/src/main/kotlin/org/openmhealth/schema/domain/ieee/DescriptiveStatistic.kt
package org.openmhealth.schema.domain.ieee

import org.openmhealth.schema.domain.IEEE_NAMESPACE
import org.openmhealth.schema.domain.SchemaEnumValue
import org.openmhealth.schema.domain.SchemaId
import org.openmhealth.schema.domain.SchemaSupport

/**
 * The descriptive statistic of a set of measurements. A measurement value can be the result of combining
 * various measurements and calculating descriptive statistics like average, maximum, minimum, etc. Additional
 * descriptive statistics will be added as the need arises. A measurement value without a descriptive statistic
 * is interpreted as being the result of an individual measurement.
 *
 * @author Emerson Farrugia
 * @version 1.0
 * @see [descriptive-statistic][https://w3id.org/ieee/ieee-1752-schema/descriptive-statistic.json]
 */
enum class DescriptiveStatistic(
    private val customSchemaValue: String? = null
) : SchemaEnumValue, SchemaSupport {

    AVERAGE,
    COUNT,
    MAXIMUM,
    MEDIAN,
    MINIMUM,
    STANDARD_DEVIATION,
    SUM,
    VARIANCE,
    TWENTIETH_PERCENTILE("20th percentile"),
    EIGHTIETH_PERCENTILE("80th percentile"),
    LOWER_QUARTILE,
    UPPER_QUARTILE,
    QUARTILE_DEVIATION,
    FIRST_QUINTILE("1st quintile"),
    SECOND_QUINTILE("2nd quintile"),
    THIRD_QUINTILE("3rd quintile"),
    FOURTH_QUINTILE("4th quintile");

    override val schemaValue: String
        get() = customSchemaValue ?: name.lowercase().replace("_", " ")

    override val schemaId: SchemaId
        get() = SchemaId(IEEE_NAMESPACE.value, "descriptive-statistic", "1.0")
}
